File: app/src/androidTest/java/com/juawapps/newsspread/db/ArticleDaoTest.java
package com.juawapps.newsspread.db;

import android.support.test.filters.MediumTest;
import android.support.test.runner.AndroidJUnit4;

import com.juawapps.newsspread.data.db.ArticleDao;
import com.juawapps.newsspread.data.objects.Article;
import com.juawapps.newsspread.util.AppDatabaseResource;
import com.juawapps.newsspread.util.LiveDataTestUtil;
import com.juawapps.newsspread.util.ObjectCreatorsUtil;

import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.CoreMatchers.notNullValue;
import static org.hamcrest.MatcherAssert.assertThat;

/**
 * Article Dao tests.
 * TODO: add tests to test date field.
 */

@RunWith(AndroidJUnit4.class)
public class ArticleDaoTest {

    @Rule
    public AppDatabaseResource mAppDatabaseResource = new AppDatabaseResource();

    @Test
    @MediumTest
    public void insertAndRead() throws Exception {
        String articleTitle = "Penguins use geisers to go to space!";

        ArticleDao articleInsert = mAppDatabaseResource.getDb().articleDao();

        // Insert article
        Article newArticle = ObjectCreatorsUtil.createArticle(articleTitle);
        articleInsert.insertArticles(new ArrayList<>(Collections.singletonList(newArticle)));

        // Read article
        List<Article> articleList = LiveDataTestUtil.getValue(articleInsert.loadAllArticles());

        assertThat(articleList.size(), is(1));
        Article articleRead = articleList.get(0);
        assertThat(articleRead, notNullValue());
        assertThat(articleRead.getTitle(), is(articleTitle));
        assertThat(articleRead.getSource(), notNullValue());
        assertThat(articleRead.getSource().getId(), is(ObjectCreatorsUtil.DEFAULT_SOURCE_ID));
        assertThat(articleRead.getSource().getName(), is(ObjectCreatorsUtil.DEFAULT_SOURCE_NAME));
        assertThat(articleRead.getAuthor(), is(ObjectCreatorsUtil.DEFAULT_ARTICLE_AUTHOR));
        assertThat(articleRead.getCategory(), is(ObjectCreatorsUtil.DEFAULT_ARTICLE_CATEGORY));
        assertThat(articleRead.getDescription(), is(ObjectCreatorsUtil.DEFAULT_ARTICLE_DESCRIPTION));
        assertThat(articleRead.getUrl(), is(ObjectCreatorsUtil.DEFAULT_ARTICLE_URL));
        assertThat(articleRead.getUrlToImage(), is(ObjectCreatorsUtil.DEFAULT_ARTICLE_URL_TO_IMAGE));
        assertThat(articleRead.getPublishedAt(), is(ObjectCreatorsUtil.DEFAULT_ARTICLE_PUBLISHED_AT));
    }

    @Test
    @MediumTest
    public void insertDuplicate() throws Exception {
        String articleTitle = "Vodka price drops!";

        ArticleDao articleInsert = mAppDatabaseResource.getDb().articleDao();

        // Insert article twice
        Article newArticle = ObjectCreatorsUtil.createArticle(articleTitle);
        articleInsert.insertArticles(new ArrayList<>(Collections.singletonList(newArticle)));
        articleInsert.insertArticles(new ArrayList<>(Collections.singletonList(newArticle)));

        // Read article
        List<Article> articleList = LiveDataTestUtil.getValue(articleInsert.loadAllArticles());

        assertThat(articleList.size(), is(1));
    }

    @Test
    @MediumTest
    public void loadCategory() throws Exception {
        String techCategory = "tech";
        String worldCategory = "world";

        String articleTechTitle = "Android is the best.";
        String articleWorldTitle = " ";

        ArticleDao articleInsert = mAppDatabaseResource.getDb().articleDao();

        // Insert articles
        Article techArticle = ObjectCreatorsUtil.createArticle(articleTechTitle, techCategory);
        Article worldArticle = ObjectCreatorsUtil.createArticle(articleWorldTitle, worldCategory);
        articleInsert.insertArticles(new ArrayList<>(Collections.singletonList(techArticle)));
        articleInsert.insertArticles(new ArrayList<>(Collections.singletonList(worldArticle)));

        // Read articles
        List<Article> articleTechList =
                LiveDataTestUtil.getValue(articleInsert.loadArticlesByCategory(techCategory));
        List<Article> articleWorldList =
                LiveDataTestUtil.getValue(articleInsert.loadArticlesByCategory(worldCategory));

        assertThat(articleTechList.size(), is(1));
        assertThat(articleWorldList.size(), is(1));

        assertThat(articleTechList.get(0).getTitle(), is(articleTechTitle));
        assertThat(articleWorldList.get(0).getTitle(), is(articleWorldTitle));
    }
}

File: app/src/androidTest/java/com/juawapps/newsspread/db/ArticleDaoTest.kt
package com.juawapps.newsspread.db

import android.support.test.filters.MediumTest
import android.support.test.runner.AndroidJUnit4
import com.juawapps.newsspread.util.*

import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

import java.util.ArrayList

import org.hamcrest.CoreMatchers.`is`
import org.hamcrest.CoreMatchers.notNullValue
import org.hamcrest.MatcherAssert.assertThat

/**
 * Article Dao tests.
 * TODO: add tests to test date field.
 */

@RunWith(AndroidJUnit4::class)
class ArticleDaoTest {

    @JvmField
    @Rule
    var mAppDatabaseResource = AppDatabaseResource()

    @Test
    @MediumTest
    @Throws(Exception::class)
    fun insertAndRead() {
        val articleTitle = "Penguins use geisers to go to space!"

        val articleInsert = mAppDatabaseResource.db.articleDao()

        // Insert article
        val newArticle = ObjectCreatorsUtil.createArticle(articleTitle)
        articleInsert.insertArticles(ArrayList(listOf(newArticle)))

        // Read article
        val articleList = LiveDataTestUtil.getValue(articleInsert.loadAllArticles())

        assertThat(articleList.size, `is`(1))
        val articleRead = articleList[0]
        assertThat(articleRead, notNullValue())
        assertThat(articleRead.title, `is`(articleTitle))
        assertThat(articleRead.source, notNullValue())
        assertThat(articleRead.source?.id, `is`(ObjectCreatorsUtil.DEFAULT_SOURCE_ID))
        assertThat(articleRead.source?.name, `is`(ObjectCreatorsUtil.DEFAULT_SOURCE_NAME))
        assertThat(articleRead.author, `is`(ObjectCreatorsUtil.DEFAULT_ARTICLE_AUTHOR))
        assertThat(articleRead.category, `is`(ObjectCreatorsUtil.DEFAULT_ARTICLE_CATEGORY))
        assertThat(articleRead.description, `is`(ObjectCreatorsUtil.DEFAULT_ARTICLE_DESCRIPTION))
        assertThat(articleRead.url, `is`(ObjectCreatorsUtil.DEFAULT_ARTICLE_URL))
        assertThat(articleRead.urlToImage, `is`(ObjectCreatorsUtil.DEFAULT_ARTICLE_URL_TO_IMAGE))
        assertThat(articleRead.publishedAt, `is`(ObjectCreatorsUtil.DEFAULT_ARTICLE_PUBLISHED_AT))
    }

    @Test
    @MediumTest
    @Throws(Exception::class)
    fun insertDuplicate() {
        val articleTitle = "Vodka price drops!"

        val articleInsert = mAppDatabaseResource.db.articleDao()

        // Insert article twice
        val newArticle = ObjectCreatorsUtil.createArticle(articleTitle)
        articleInsert.insertArticles(listOf(newArticle))
        articleInsert.insertArticles(listOf(newArticle))

        // Read article
        val articleList = LiveDataTestUtil.getValue(articleInsert.loadAllArticles())

        assertThat(articleList.size, `is`(1))
    }

    @Test
    @MediumTest
    @Throws(Exception::class)
    fun loadCategory() {
        val techCategory = "tech"
        val worldCategory = "world"

        val articleTechTitle = "Android is the best."
        val articleWorldTitle = " "

        val articleInsert = mAppDatabaseResource.db.articleDao()

        // Insert articles
        val techArticle = ObjectCreatorsUtil.createArticle(articleTechTitle, techCategory)
        val worldArticle = ObjectCreatorsUtil.createArticle(articleWorldTitle, worldCategory)
        articleInsert.insertArticles(ArrayList(listOf(techArticle)))
        articleInsert.insertArticles(ArrayList(listOf(worldArticle)))

        // Read articles
        val articleTechList = LiveDataTestUtil.getValue(articleInsert.loadArticlesByCategory(techCategory))
        val articleWorldList = LiveDataTestUtil.getValue(articleInsert.loadArticlesByCategory(worldCategory))

        assertThat(articleTechList.size, `is`(1))
        assertThat(articleWorldList.size, `is`(1))

        assertThat(articleTechList[0].title, `is`(articleTechTitle))
        assertThat(articleWorldList[0].title, `is`(articleWorldTitle))
    }
}
